csharp/src/Ice/IncomingFrame.cs

```diff
@@ -0,0 +1,96 @@
+//
+// Copyright (c) ZeroC, Inc. All rights reserved.
+//
+
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.IO.Compression;
+
+namespace ZeroC.Ice
+{
+    /// <summary>Base class for incoming frames.</summary>
+    public abstract class IncomingFrame
+    {
+        /// <summary>The encoding of the frame payload.</summary>
+        public abstract Encoding Encoding { get; }
+        /// <summary>The payload of this request frame. The bytes inside the payload should not be written to;
+        /// they are writable because of the <see cref="System.Net.Sockets.Socket"/> methods for sending.</summary>
+        // TODO: describe how long this payload remains valid once we add memory pooling.
+        public abstract ArraySegment<byte> Payload { get; internal set; }
+
+        /// <summary>The Ice protocol of this frame.</summary>
+        public Protocol Protocol { get; }
+
+        /// <summary>The frame byte count</summary>
+        public int Size { get; private set; }
+
+        // Ice2 compression status, is the first byte of the encaps payload
+        internal byte CompressionStatus
+        {
+            get
+            {
+                Debug.Assert(Encoding == Encoding.V2_0);
+                int sizeLength = 1 << (Payload[0] & 0x03);
```

Comment:
This assumes the payload is an encapsulation. That's always true when protocol is ice2, but not when protocol is ice1 - in particular for an ice1 incoming response frame, the first byte is always ReplyStatus.

Reply:
I see for a response we can use something like

```
public override bool HasCompressedPayload =>
            Encoding == Encoding.V2_0 &&
            (Protocol == Protocol.Ice2 || (Protocol == Protocol.Ice1 && (ReplyStatus == ReplyStatus.OK ||
                                                                         ReplyStatus == ReplyStatus.UserException))) &&
            Payload[Payload.AsReadOnlySpan().ReadSize(Protocol.GetEncoding()).SizeLength + 2] != 0;
```

It will be similar for a request but without the ReplyStatus

Reply:
Yes, but I think it would be simpler to make `HasCompressedPayload` an auto-implemented bool property.

Reply:
Added auto-implemented property for `HasCompressedPayload `